src/android/aidl/birthday_service/src/lib.rs

```diff
@@ -26,4 +30,61 @@ impl IBirthdayService for BirthdayService {
     fn wishHappyBirthday(&self, name: &str, years: i32) -> binder::Result<String> {
         Ok(format!("Happy Birthday {name}, congratulations with the {years} years!"))
     }
+    // ANCHOR_END: IBirthdayService
+
+    fn wishWithInfo(&self, info: &BirthdayInfo) -> binder::Result<String> {
+        Ok(format!(
+            "Happy Birthday {}, congratulations with the {} years!",
+            info.name, info.years,
+        ))
+    }
+
+    fn wishWithProvider(
+        &self,
+        provider: &Strong<dyn IBirthdayInfoProvider>,
+    ) -> binder::Result<String> {
+        Ok(format!(
+            "Happy Birthday {}, congratulations with the {} years!",
+            provider.name()?,
+            provider.years()?,
+        ))
+    }
+
+    fn wishWithErasedProvider(
+        &self,
+        provider: &SpIBinder,
+    ) -> binder::Result<String> {
+        use binder::binder_impl::Proxy;
+
+        // Convert the `IBinder` to a concrete interface.
+        let provider = BpBirthdayInfoProvider::from_binder(provider.clone())?;
+
+        Ok(format!(
+            "Happy Birthday {}, congratulations with the {} years!",
+            provider.name()?,
+            provider.years()?,
+        ))
+    }
+
+    // ANCHOR: wishFromFile
+    fn wishFromFile(
+        &self,
+        info_file: &ParcelFileDescriptor,
+    ) -> binder::Result<String> {
+        let mut info_file = info_file
+            .as_ref()
+            .try_clone()
```

Comment:
Might be useful or interesting to explain what's going on here (either verbally, in the notes, or in a code comment here): `ParcelFileDescriptor` wraps a `std::os::fd::OwnedFd` but we want a `File`. We need to convert one into the other.

Reply:
This code isn't included in the slides, but I'll add a comment explaining the conversion.